src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,168 @@
+        }
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        if (index < 0 || index > size) {
+            throw new IndexOutOfBoundsException();
+        }
+
+        if (index == size) {
+            add(value);
+            return;
+        }
+
+        Node<T> newNode = new Node<>(value);
+        if (index == 0) {
+            newNode.next = first;
+            first.prev = newNode;
+            first = newNode;
+        } else {
+            Node<T> current = getNodeAtIndex(index);
+            newNode.prev = current.prev;
+            newNode.next = current;
+            current.prev.next = newNode;
+            current.prev = newNode;
+        }
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T value : list) {
+            add(value);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        return getNodeAtIndex(index).data;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndex(index);
+        Node<T> node = getNodeAtIndex(index);
+        T oldValue = node.data;
+        node.data = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index);
+        Node<T> current = getNodeAtIndex(index);
+
+        if (current == first) {
+            first = current.next;
+            if (first != null) {
+                first.prev = null;
+            }
+        } else if (current == last) {
+            last = current.prev;
+            if (last != null) {
+                last.next = null;
+            }
+        } else {
+            unlink(current); // Reuse the unlink logic
+        }
+        size--;
+        return current.data;
     }
 
     @Override
     public boolean remove(T object) {
+        Node<T> current = first;
+        while (current != null) {
+            if (object == null ? current.data == null : object.equals(current.data)) {
+                unlink(current);
+                size--;
+                return true;
+            }
+            current = current.next;
+        }
         return false;
     }
 
+    private void unlink(Node<T> node) {
+        Node<T> prevNode = node.prev;
+        Node<T> nextNode = node.next;
+
+        if (prevNode != null) {
+            prevNode.next = nextNode;
+        } else {
+            first = nextNode;
+        }
+
+        if (nextNode != null) {
+            nextNode.prev = prevNode;
+        } else {
+            last = prevNode;
+        }
+        node.prev = null;
+        node.next = null;
+    }
+
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
     }
+
+    private Node<T> getNodeAtIndex(int index) {
+        if (index < size / 2) {
```

Comment:
let's `checkIndex()` inside `getNodeAtIndex(...)` method to avoid code duplication. 
Actually, it is a part of `getNodeAtIndex(...)` logic, isn't it?